toolchain/check/context.cpp

```diff
@@ -187,6 +187,82 @@ auto Context::DiagnoseNameNotFound(SemIRLoc loc, SemIR::NameId name_id)
   emitter_->Emit(loc, NameNotFound, name_id);
 }
 
+auto Context::DiagnoseInvalidAccess(
+    Parse::NodeId node_id, SemIR::InstId member_id, SemIR::NameId name_id,
+    SemIR::AccessKind access_kind, bool is_parent_access,
+    std::optional<AccessInfo> access_info) -> void {
+  if (!access_info) {
+    return;
+  }
+
+  auto inst = insts().Get(access_info->base_id);
+  auto inst_type_id = inst.type_id();
+
+  if (auto class_type = types().TryGetAs<SemIR::ClassType>(inst_type_id)) {
+    auto class_inst = classes().Get(class_type->class_id);
+    auto is_access_from_instance_method = access_info->allow_protected_access &&
+                                          access_info->allow_private_access;
+
+    CARBON_DCHECK(class_inst.is_defined());
```

Comment:
```suggestion
    CARBON_CHECK(class_inst.is_defined());
```

We're only using DCHECK where a small performance cost still significantly affects performance. When diagnosing invalid code, there isn't much reason to avoid CHECK -- invalid code is expected to perform worse for compilation.